### ISSUE TITLE
Implement `uncount()`

### ISSUE THREAD

**Issue by markfairbanks** [tidyverse/dtplyr#352]
``` r
library(tidyr)

df <- tibble(x = c("a", "b"), n = c(1, 2))

df %>%
  uncount(n)
#> # A tibble: 3 × 1
#>   x    
#>   <chr>
#> 1 a    
#> 2 b    
#> 3 b
```

**Comment by markfairbanks** [comment]
Looks like we would need to make `uncount()` a generic in tidyr first.

**Comment by mgirlich** [comment]
There is already a tidyr issue https://github.com/tidyverse/tidyr/issues/1071 for this. In dbplyr we simply added `dbplyr_uncount()` for now...

**Comment by markfairbanks** [comment]
Should we add `dtplyr_uncount()`? Or maybe we should wait for https://github.com/tidyverse/tidyr/pull/1101 to be merged?

**Comment by mgirlich** [comment]
It might take some time for https://github.com/tidyverse/tidyr/pull/1101 to be merged (the interface of most of these functions is not completely stable yet). Therefore, I created a separate PR https://github.com/tidyverse/tidyr/pull/1358 for `uncount()`. This might be merged faster.

**Comment by markfairbanks** [comment]
Here's an initial implementation that I think covers everything. Once https://github.com/tidyverse/tidyr/pull/1358 is merged we can create this as a `uncount.dtplyr_step` method.

``` r
library(data.table)
library(dtplyr)
library(dplyr, warn.conflicts = FALSE)

dtplyr_uncount <- function(data, weights, ..., .remove = TRUE, .id = NULL) {
  weights <- enquo(weights)
  
  out <- data
  
  if (!is.null(.id)) {
    out <- mutate(out, !!.id := row_number())
  }
  
  out <- slice(out, rep(1:.N, !!weights))
  
  if (.remove) {
    out <- select(out, -!!weights)
  }
  
  out
}

df <- data.table(x = c("a", "b"), n = c(1, 2))

df %>%
  dtplyr_uncount(n)
#> Source: local data table [3 x 1]
#> Call:   `_DT1`[rep(1:.N, n)[between(rep(1:.N, n), -.N, .N)], .(x)]
#> 
#>   x    
#>   <chr>
#> 1 a    
#> 2 b    
#> 3 b    
#> 
#> # Use as.data.table()/as.data.frame()/as_tibble() to access results

df %>%
  dtplyr_uncount(n, .id = "id")
#> Source: local data table [3 x 2]
#> Call:   copy(`_DT2`)[, `:=`(id = seq_len(.N))][rep(1:.N, n)[between(rep(1:.N, 
#>     n), -.N, .N)], .(x, id)]
#> 
#>   x        id
#>   <chr> <int>
#> 1 a         1
#> 2 b         2
#> 3 b         2
#> 
#> # Use as.data.table()/as.data.frame()/as_tibble() to access results
```

**Comment by markfairbanks** [comment]
I think might be a better way - it avoids the `pull()` issue I mentioned above.

``` r
library(data.table)
library(dtplyr)
library(dplyr, warn.conflicts = FALSE)
library(tidyr, warn.conflicts = FALSE)
library(purrr)

step_subset_j <- dtplyr:::step_subset_j

dtplyr_uncount <- function(data, weights, ..., .remove = TRUE, .id = NULL) {
  weights <- enexpr(weights)
  
  if (is.null(.id)) {
    out <- slice(data, rep(1:.N, !!weights))
    if (.remove) {
      out <- select(out, -!!weights)
    }
  } else {
    vars_names <- data$vars
    if (.remove) {
      vars_names <- setdiff(vars_names, as_name(weights))
    }
    vars <- map(syms(vars_names), ~ expr(rep(!!.x, !!weights)))
    names(vars) <- vars_names
    
    vars <- append(vars, exprs(!!.id := sequence(!!weights)))

    out <- step_subset_j(
      data,
      vars = union(data$groups, vars_names),
      j = call2(".", !!!vars)
    )
  }
  
  out
}

df <- lazy_dt(data.table(x = c("a", "b"), n = c(1, 2)))

df %>%
  dtplyr_uncount(n, .remove = FALSE)
#> Source: local data table [3 x 2]
#> Call:   `_DT1`[rep(1:.N, n)[between(rep(1:.N, n), -.N, .N)]]
#> 
#>   x         n
#>   <chr> <dbl>
#> 1 a         1
#> 2 b         2
#> 3 b         2
#> 
#> # Use as.data.table()/as.data.frame()/as_tibble() to access results

df %>%
  dtplyr_uncount(n, .id = "id", .remove = FALSE)
#> Source: local data table [3 x 3]
#> Call:   `_DT1`[, .(x = rep(x, n), n = rep(n, n), id = sequence(n))]
#> 
#>   x         n    id
#>   <chr> <dbl> <int>
#> 1 a         1     1
#> 2 b         2     1
#> 3 b         2     2
#> 
#> # Use as.data.table()/as.data.frame()/as_tibble() to access results
```

**Comment by eutwt** [comment]
Would it be faster to just always use the second branch of your function, where you `rep` each column? (obviously it is in the example below, but I haven't thought about the other possible scenarios)

``` r
devtools::load_all('/Users/mbp/Documents/GitHub/dtp')
#> ℹ Loading dtplyr
#> Warning: package 'dplyr' was built under R version 4.1.2


# dtplyr_uncount <- [as above]

dtplyr_uncount2 <- function(data, weights, ..., .remove = TRUE, .id = NULL) {
  weights <- quo_squash(enquo(weights))
  
  groups <- group_vars(data)
  has_groups <- length(groups) > 0
  if (has_groups) {
    data <- ungroup(data)
  }

  vars_names <- data$vars
  if (.remove) {
    vars_names <- setdiff(vars_names, as_name(weights))
  }
  vars <- map(syms(vars_names), ~ expr(rep(!!.x, !!weights)))
  names(vars) <- vars_names
  
  if (!is.null(.id)) {
    vars <- append(vars, exprs(!!.id := sequence(!!weights)))
  }

  out <- step_subset_j(
    data,
    vars = vars_names,
    j = call2(".", !!!vars)
  )
  
  if (has_groups) {
    out <- group_by(out, !!!syms(groups))
  }
  
  out
}

df <- data.table(x = c("a", "b"), n = c(1, 2)) %>% 
  mutate(n = c(1, 2)*1e6) %>% 
  lazy_dt()


library(bench)

mark(
  a =  df %>% dtplyr_uncount(n) %>% collect,
  b =  df %>% dtplyr_uncount2(n) %>% collect)
#> Warning: Some expressions had a GC in every iteration; so filtering is disabled.
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 a            55.1ms   60.4ms      13.9    69.4MB     21.8
#> 2 b            19.2ms   31.7ms      25.7    46.1MB     19.8
```

<sup>Created on 2022-05-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by markfairbanks** [comment]
> Would it be faster to just always use the second branch of your function

Looks like it. I hadn’t gotten around to testing it yet. I’m a bit surprised to be honest - I would have assumed the simple slice/select would be much more efficient.
